File: ts/project/.sdk/tm/lua/utility/make_response.lua
-- ProjectName SDK utility: make_response

local function make_response_util(ctx)
  if ctx.out["response"] ~= nil then
    return ctx.out["response"], nil
  end

  local utility = ctx.utility
  local spec = ctx.spec
  local result = ctx.result
  local response = ctx.response

  if spec == nil then
    return nil, ctx:make_error("response_no_spec",
      "Expected context spec property to be defined.")
  end
  if response == nil then
    return nil, ctx:make_error("response_no_response",
      "Expected context response property to be defined.")
  end
  if result == nil then
    return nil, ctx:make_error("response_no_result",
      "Expected context result property to be defined.")
  end

  spec.step = "response"

  utility.result_basic(ctx)
  utility.result_headers(ctx)
  utility.result_body(ctx)
  utility.transform_response(ctx)

  if result.err == nil then
    result.ok = true
  end

  if ctx.ctrl.explain ~= nil then
    ctx.ctrl.explain["result"] = result
  end

  return response, nil
end

return make_response_util

File: ts/project/.sdk/tm/csharp/utility/MakeResponse.cs
// ProjectName SDK utility: makeResponse - shape the transport response
// into the result.

namespace ProjectNameSdk.Util;

public static partial class SdkUtility
{
    internal static Response MakeResponseUtil(Context ctx)
    {
        if (ctx.Out.TryGetValue("response", out var outResp) && outResp is Response cached)
        {
            return cached;
        }

        var utility = ctx.Utility!;
        var spec = ctx.Spec;
        var result = ctx.Result;
        var response = ctx.Response;

        if (spec == null)
        {
            throw ctx.MakeError("response_no_spec",
                "Expected context spec property to be defined.");
        }
        if (response == null)
        {
            throw ctx.MakeError("response_no_response",
                "Expected context response property to be defined.");
        }
        if (result == null)
        {
            throw ctx.MakeError("response_no_result",
                "Expected context result property to be defined.");
        }

        spec.Step = "response";

        utility.ResultBasic(ctx);
        utility.ResultHeaders(ctx);
        utility.ResultBody(ctx);
        utility.TransformResponse(ctx);

        if (result.Err == null)
        {
            result.Ok = true;
        }

        if (ctx.Ctrl.Explain != null)
        {
            ctx.Ctrl.Explain["result"] = result;
        }

        return response;
    }
}
